NIOTests.ChannelTests testCloseInReadTriggeredByDrainingTheReceiveBufferBecauseOfWriteError Flaky

Happened in macOS CI

```
03:03:42 - Test Case '-[NIOTests.ChannelTests testCloseInput]' passed (0.004 seconds).
03:03:42 - Test Case '-[NIOTests.ChannelTests testCloseInReadTriggeredByDrainingTheReceiveBufferBecauseOfWriteError]' started.
03:03:42 - /private/tmp/task-sandbox/5069235300211106632.xL7gDH/0~mMnRRsM8DUUqjaxSgBUgebeIn3-pYHuzNYCRubOCouO9LEH3qda9nvndcvoBXLVE9U1nGrxmpEzZbja73llEiw==/SRCROOT/Tests/NIOTests/ChannelTests.swift:2530: error: -[NIOTests.ChannelTests testCloseInReadTriggeredByDrainingTheReceiveBufferBecauseOfWriteError] : XCTAssertEqual failed: ("Optional(NIO.ChannelError.ioOnClosedChannel)") is not equal to ("nil") - 
03:03:42 - Test Case '-[NIOTests.ChannelTests testCloseInReadTriggeredByDrainingTheReceiveBufferBecauseOfWriteError]' failed (0.006 seconds).
```

--- a/Sources/NIO/System.swift
+++ b/Sources/NIO/System.swift
@@ -60,6 +60,7 @@ private let sysOpenWithMode: @convention(c) (UnsafePointer<CChar>, CInt, mode_t)
 private let sysWrite: @convention(c) (CInt, UnsafeRawPointer?, CLong) -> CLong = write
 private let sysRead: @convention(c) (CInt, UnsafeMutableRawPointer?, CLong) -> CLong = read
 private let sysLseek: @convention(c) (CInt, off_t, CInt) -> off_t = lseek
+private let sysPoll: @convention(c) (UnsafeMutablePointer<pollfd>?, nfds_t, Int32) -> CInt = poll
 #if os(Android)
 func sysRecvFrom_wrapper(sockfd: CInt, buf: UnsafeMutableRawPointer, len: CLong, flags: CInt, src_addr: UnsafeMutablePointer<sockaddr>, addrlen: UnsafeMutablePointer<socklen_t>) -> CLong {
     return recvfrom(sockfd, buf, len, flags, src_addr, addrlen) // src_addr is 'UnsafeMutablePointer', but it need to be 'UnsafePointer'
@@ -486,6 +487,13 @@ internal enum Posix {
     public static func if_nametoindex(_ name: UnsafePointer<CChar>?) throws -> CUnsignedInt {
         return try wrapSyscall {
             sysIfNameToIndex(name)
+        }
+    }
+
+    @inline(never)
+    public static func poll(fds: UnsafeMutablePointer<pollfd>, nfds: nfds_t, timeout: CInt) throws -> CInt {
+        return try wrapSyscall {
+            sysPoll(fds, nfds, timeout)
         }
     }
 }

--- a/Tests/NIOTests/ChannelTests.swift
+++ b/Tests/NIOTests/ChannelTests.swift
@@ -2506,6 +2506,18 @@ public class ChannelTests: XCTestCase {
                 self.allDonePromise = allDonePromise
             }
 
+            private func veryNasty_blockUntilReadBufferIsNonEmpty(channel: Channel) throws {
+                struct ThisIsNotASocketChannelError: Error {}
+                guard let channel = channel as? SocketChannel else {
+                    throw ThisIsNotASocketChannelError()
+                }
+                try channel.socket.withUnsafeFileDescriptor { fd in
+                    var pollFd: pollfd = .init(fd: fd, events: Int16(POLLIN), revents: 0)
+                    let nfds = try Posix.poll(fds: &pollFd, nfds: 1, timeout: -1)
+                    XCTAssertEqual(1, nfds)
+                }
+            }
+
             func channelActive(ctx: ChannelHandlerContext) {
                 XCTAssert(serverChannel.eventLoop === ctx.eventLoop)
                 self.serverChannel.whenSuccess { serverChannel in
@@ -2523,11 +2535,22 @@ public class ChannelTests: XCTestCase {
                         // let's trigger the write error
                         var buffer = ctx.channel.allocator.buffer(capacity: 16)
                         buffer.write(staticString: "THIS WILL FAIL ANYWAY")
+
+                        // this needs to be in a function as otherwise the Swift compiler believes this is throwing
+                        func workaroundSR487() {
+                            // this test only tests the correct condition if the bytes sent from the other side have already
+                            // arrived at the time the write fails. So this is a hack that makes sure they do have arrived.
+                            // (https://github.com/apple/swift-nio/issues/657)
+                            XCTAssertNoThrow(try self.veryNasty_blockUntilReadBufferIsNonEmpty(channel: ctx.channel))
+                        }
+                        workaroundSR487()
+
                         return ctx.writeAndFlush(self.wrapOutboundOut(buffer))
                     }.map {
                         XCTFail("this should have failed")
                     }.whenFailure { error in
-                        XCTAssertEqual(ChannelError.ioOnClosedChannel, error as? ChannelError)
+                        XCTAssertEqual(ChannelError.ioOnClosedChannel, error as? ChannelError,
+                                       "unexpected error: \(error)")
                         XCTAssertTrue(inSameStackFrame)
                         self.allDonePromise.succeed(())
                     }

